proposals/0012-introduce-reactNativeMetadata.md

```diff
@@ -0,0 +1,343 @@
+While an app project can contain native modules and components, it's **not** possible for a app to "consume" another app as a library as this would break the native build system invariants (i.e. a single `AndroidManifest.xml` and `Info.plist` file per app, etc.). If an app projects wishes to distribute some of its native modules or components, they need to be extracted into a library project.
+
+We're proposing to declare if a project is either a library or an app by adding the `type` field (as `"app"` or `"library"`) to the `reactNativeManifest` section of the `package.json` file.
+
+```json
+{
+  "reactNativeManifest": {
+    "type": "library"
+  }
+}
+```
+
+### New Architecture support
+
+One of the primary driver of this proposal is the [New React Native Architecture support](https://reactnative.dev/docs/the-new-architecture/landing-page). Specifically we believe that the ecosystem is currently affected by:
+* Lack of a declarative way to define if a library is compatible with New Architecture
+  * As described in the [rationale](#rationale), tooling needs to infer this information from the presence of the `codegenConfig` field in the `package.json` of a library. Libraries might decide to don't use codegen and still be compatible with New Architecture, so this method is not exhaustive.
+  * The alternative at this stage is to inspect the code and check the API callsites if they are New Architecture or Old Architecture compatible.
+* Lack of a declarative way to enable New Architecture on both platforms
+  * Currently New Architecture is enabled by a Gradle Property `newArchEnabled` on Android in the `gradle.properties` file and by invokign `RCT_NEW_ARCH_ENABLED pod install` on iOS.
+  * As of today, there is no way to enable New Architecture for an app project for both platforms
+  * Moreover, the file vs environment variable leads to a scenario where you can't know if a app supports New Architecture or not by inspecting the code (as New Architecture support is known at  `pod install` time and is not codified in the codebase)
+
+Therefore we propose to add the `newArch` section to the `reactNativeManifest.capabilities` of **both apps and libraries** with the following semantic:
+* For Apps: `newArch.enabled==true` implies that the app wants to use the New Architecture.
+* For Libraries: `newArch.enabled==true` implies that the library is compatible with the New Architecture.
+
+Tools can be built on top of this information to check that a app with `newArch.enabled==true` is not accepting libraries with `newArch.enabled==false` and warning against library that don't have the key specified.
+
+The setup would look as follows for both apps and libraries:
+
+```json
+{
+  "reactNativeManifest": {
+    "capabilities": {
+      "newArch": {
+        "enabled": true
+      }
+    }
+  }
+}
+```
+
+This section will allows also for split configuration between platforms:
```

Comment:
Shouldn't we specify the precedence rules? Because a setup like:

```json
{
  "reactNativeManifest": {
    "capabilities": {
      "android": {
      "capabilities": {
        "newArch": {
          "enabled": true
        }
      }
    }
  }
}
```

Is invalid. What is the definition for iOS?

Secondly: do we want to allow for Libraries to say that they supports the new architecture only for one of the two platforms or not?